src/librustc/lint/internal.rs

```diff
@@ -225,3 +225,31 @@ fn gen_args(segment: &PathSegment) -> String {
 
     String::new()
 }
+
+declare_lint! {
+    pub LINT_PASS_IMPL_WITHOUT_MACRO,
+    Allow,
+    "`impl LintPass` without the `declare_lint_pass!` or `impl_lint_pass!` macros"
+}
+
+declare_lint_pass!(LintPassImpl => [LINT_PASS_IMPL_WITHOUT_MACRO]);
+
+impl EarlyLintPass for LintPassImpl {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if let ItemKind::Impl(_, _, _, _, Some(lint_pass), _, _) = &item.node {
+            if !lint_pass.path.span.ctxt().outer_expn_info().is_some() {
```

Comment:
I guess any macro is better than none, checking for `declare_lint_pass!` or `impl_lint_pass!` specifically is overkill?

Reply:
Didn't thought of specifically checking for these macros. I'll change that (since it's not much work), but don't think that there will be more lint findings.